### DISCUSSION TITLE
How does auto-update work?

### DISCUSSION THREAD

**Discussion by se7entynine** [moghtech/komodo#238] [Q&A]
Basically the title - how does auto update work? I enalbed both options but it doesn't update the stack and a new update is visible through portainer. 

Is there an indicator I have to set or is updating a stack only manually possible?

**Comment by kingp0dd** [comment]
Would also like to ask. I don't have auto update enabled but auto poll is. how do i know if theres a new image update available? from docs, i think it should be in Alerts? but i haven't seen one notification from there yet

**Comment by mbecker20** [comment]
The auto poll for update works by using docker pull with Deployments, or docker compose pull in the case of Stacks. It will run this on an interval set by resource_poll_interval in the config: https://github.com/mbecker20/komodo/blob/main/config/core.config.toml#L373. By default this is every 5 minutes.

If there is a newer image available, then the newer image will be pulled, and Komodo will know that the hash between running image and latest image does not match. This is how it checks for update available, through combination of docker pull and hash equality check. If this is detected, Komodo will send the information on the resource.info.update_available field for Deployments: https://docs.rs/komodo_client/latest/komodo_client/entities/deployment/struct.DeploymentListItemInfo.html#structfield.update_available. For Stacks, it must be done at the service level, eg stack.info.services[0].update_available: https://docs.rs/komodo_client/latest/komodo_client/entities/stack/struct.StackListItemInfo.html#structfield.services.

It will also send the Alert when it sees the hash mismatch:
https://docs.rs/komodo_client/latest/komodo_client/entities/alert/enum.AlertData.html#variant.DeploymentImageUpdateAvailable and https://docs.rs/komodo_client/latest/komodo_client/entities/alert/enum.AlertData.html#variant.StackImageUpdateAvailable.

If you enable auto_update, a redeploy will be triggered when it finds hash mismatch for the images after docker pull.

Let me know if you have any more questions!

**Reply by kingp0dd** [reply]
Hi! I posted this earlier https://github.com/mbecker20/komodo/issues/240 I'm not sure if the Stack Alert was intended to be marked as Resolved even if the hash equality check does not pass yet.

**Reply by hannibal85** [reply]
> If there is a newer image available, then the newer image will be pulled, and Komodo will know that the hash between running image and latest image does not match. This is how it checks for update available, through combination of docker pull and hash equality check. If this is detected, Komodo will send the information on the resource.info.update_available field for Deployments: https://docs.rs/komodo_client/latest/komodo_client/entities/deployment/struct.DeploymentListItemInfo.html#structfield.update_available. For Stacks, it must be done at the service level, eg stack.info.services[0].update_available: https://docs.rs/komodo_client/latest/komodo_client/entities/stack/struct.StackListItemInfo.html#structfield.services.

Sorry for bringing this old topic up again @mbecker20 
But my questions seems to fit quite good here.

Short info on my setup
- everything deployed via stacks (docker-compose.yml)
- integrated the existing stacks via "Files on server" mode to Komodo
- activated "poll for updates" ("auto update" disabled)

I'm struggling with the update notifications. I'm using in my compose files always almost tagged image versions instead of "latest" to keep full control over the used version (e.g. `image: "ghcr.io/moghtech/komodo-core:1.17.5"`). Nevertheless I would like to get an info when an updated image is available.

What is the intended behaviour of Komodo here? Should this work or not?
The behaviour I see is that it doesn't work. Only if the digest of the same tagged version was updated for whatever reason I see the "update available" information. For all other images where there is a new tag available but no change for the currently used tag I get no info.

Would be really great if this would work also for tagged version so I could retire my WUD containers and have all information at one place in Komodo.

**Reply by mbecker20** [reply]
Currently it will behave as you say. Since it only checks for newer image with same tag, it won't see there's a newer tag. It would be a nice feature to do so, but also pretty complicated to implement.

Currently users storing compose files in git repo are able to use renovate to update hardcoded tags with a git Pull Request. If you set up some webhooks, this can be used for end to end update automation as well, with the control of hardcoded tags.

**Reply by hannibal85** [reply]
Thanks for your feedback and also the recommendation regarding renovate. Already thought about this some time ago. But for now I will stay in this case with WUD regarding about notifications which is working flawless.

**Reply by hcoohb** [reply]
I too prefer to use hardcoded tags for better control of image version (instead of latest) and I use WUD too.
I also use WUD to get notifications of updates.
Here is my question:
Given that wud has webhooks to trigger update notifications, could that be somehow used to let komodo know there is an update for an image and the new tag version ?

**Comment by Solear** [comment]
Hi! In my stacks, I can see "update available" is there a new update available. Is there a way to see all available updates of all stacks in one point? And maybe update all per one click?

**Comment by mbecker20** [comment]
@Solear I have added update available indicator in the Stack list table to more easily see an overview in 1.17 (unreleased).

In terms of updating all Stacks with available updates, you can already make a new "Action" which can do this. Go to the Actions tab and make a new one called "update_all", with this Javascript:

```js
// Collect all the Stacks with available update
const stacks_with_update = (await komodo.read("ListStacks", {})).filter(
  (stack) => !stack.info.services.every((service) => !service.update_available),
);

// Trigger deploys on all of them to perform the update
Promise.all(
  stacks_with_update.map(async (stack) => {
    await komodo.execute("DeployStack", { stack: stack.name });
    console.log("Triggered update on Stack", stack.name);
  }),
);
```

**Comment by kingp0dd** [comment]
Wow is this Action a new feature?

Is the Updates Available button now clickable on mobile?

**Comment by GentleHoneyLover** [comment]
@mbecker20 thanks for the explanation! Is there a way to control the auto-update window? E. g. right now Komodo seems to be updating a container as soon as it discovered a new image available. For some services this is not desirable — I'd rather have them updated at a specific time at night. Example: Plex container updating in the middle of a movie — I'd rather have it updated between 3 am and 5 am or something. Is it achievable?

**Comment by kingp0dd** [comment]
I Didn't know WUD can do that?
